clients/python/src/objectstore_client/client.py

```diff
@@ -26,94 +28,142 @@
 Permission = Literal["read", "write"]
 
 
-class Scope(TypedDict):
-    organization: int
-    project: NotRequired[int]
-
-
 class GetResult(NamedTuple):
     metadata: Metadata
     payload: IO[bytes]
 
 
-class ClientBuilder:
+class Usecase:
+    """
+    An identifier for a workload in Objectstore, along with defaults to use for all
+    operations within that usecase.
+
+    Usecases need to be statically defined in Objectstore's configuration.
+    Objectstore can make decisions based on the usecase. For example, choosing the most
+    suitable storage backend.
+    """
+
+    name: str
+    _default_compression: Compression
+    _default_expiration_policy: ExpirationPolicy | None
+
     def __init__(
         self,
-        objectstore_base_url: str,
-        usecase: str,
+        name: str,
+        default_compression: Compression = "zstd",
+        default_expiration_policy: ExpirationPolicy | None = None,
+    ):
+        self.name = name
+        self._default_compression = default_compression
+        self._default_expiration_policy = default_expiration_policy
+
+
+SCOPE_VALUE_ALLOWED_CHARS = set(string.ascii_letters + string.digits + "-_")
+
+
+class Scope:
+    """
+    A (possibly nested) namespace within a usecase, given as a sequence of key-value
+    pairs passed as kwargs.
+    The emtpy scope is not admitted. Order of the components matters.
+
+    The admitted characters for keys and values are: `[A-Za-z0-9_-]`.
+
+    Users are free to choose the scope structure that best suits their usecase.
+    The combination of Usecase and Scope will determine the physical path of the
+    blob in the underlying storage backend.
+
+    For most usecases, [SentryScope] should be used.
+    """
+
+    def __init__(self, **scopes: str | int | bool) -> None:
+        if len(scopes) == 0:
+            raise ValueError("At least 1 scope is needed")
+
+        parts = []
+        for key, value in scopes.items():
+            value = str(value)
+            if any(c not in SCOPE_VALUE_ALLOWED_CHARS for c in value):
+                raise ValueError(
+                    f"Invalid scope value {value}. The valid character set is: "
+                    f"{''.join(SCOPE_VALUE_ALLOWED_CHARS)}"
+                )
+
+            formatted = f"{key}.{value}"
+            parts.append(formatted)
+
+        self._str = "/".join(parts)
+
+    def __repr__(self) -> str:
+        return self._str
+
+
+class SentryScope(Scope):
```

Comment:
Originally, I was not going to add this to our client. Given that we expect virtually all use cases to have these scopes I think this makes sense though and will greatly simplify usage.